modules/host/sector.go

```diff
@@ -0,0 +1,280 @@
+	// sector when there is not enough storage remaining on the host to accept
+	// the sector.
+	//
+	// Ideally, the host will adjust pricing as the host starts to fill up, so
+	// this error should be pretty rare. Demand should drive the price up
+	// faster than the Host runs out of space, such that the host is always
+	// hovering around 95% capacity and rarely over 98% or under 90% capacity.
+	errInsufficientStorageForSector = errors.New("not enough storage remaining to accept sector")
+
+	// errNoStorage is returned when there are no storage folders, meaning the
+	// host cannot support adding a new sector.
+	errNoStorage = errors.New("cannot add sector - there is no storage folder")
+)
+
+// sectorUsage indicates how a sector is being used. Each block height
+// represents a point at which a file contract using the sector expires. File
+// contracts that use the sector multiple times will have their block height
+// appear multiple times. This data allows the host to figure out what types of
+// discounts can be applied to data that is reusing sectors. This is primarily
+// useful for file contract renewals, and really shouldn't be used otherwise.
+//
+// The StorageFolder field indicates which storage folder is housing the
+// sector.
+type sectorUsage struct {
+	StorageFolder []byte
+	Expiry        []types.BlockHeight
+}
+
+// sectorID returns the id that should be used when referring to a sector.
+// There are lots of sectors, and to minimize their footprint a reduced size
+// hash is used. Hashes are typically 256bits to provide collision resistance
+// against an attacker that is able to peform an obscene number of trials per
+// second on each of an obscene number of machines. Potential collisions for
+// sectors are limited because hosts have secret data that the attacker does
+// not know which is used to salt the transformation of a sector hash to a
+// sectorID. As a result, an attacker is limited in the number of times they
+// can try to cause a collision - one random shot every time they upload a
+// sector, and the attacker has limited ability to learn of the success of the
+// attempt. Uploads are very slow, even on fast machines there will be less
+// than 1000 per second. It is therefore safe to reduce the security from
+// 256bits to 96bits, which has a collision resistance of 2^48. A reasonable
+// upper bound for the number of sectors on a host is 2^32, corresponding with
+// 16PB of data.
+//
+// 12 bytes can be represented as a filepath using 16 base64 characters. This
+// keeps the filesize small and therefore limits the amount of load placed on
+// the filesystem when trying to manage hundreds of thousands or even tens of
+// millions of sectors in a single folder.
+func (h *Host) sectorID(sectorRootBytes []byte) []byte {
+	saltedRoot := crypto.HashAll(sectorRootBytes, h.sectorSalt)
+	id := make([]byte, 16)
+	base64.RawURLEncoding.Encode(id, saltedRoot[:12])
```

Comment:
There's a good reason I had for using bytes as input instead of a hash, but it's currently escaping me. I think it had something to do with the call to `crypto.HashAll`, if you pass a hash in it'll have a different result because you end up prepending 8 bytes of `len` when it's a slice.

hmm, struggling to remember all the details though. It might not be needed anymore, definitely wish I had dropped a comment when I made it.